sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/WatermarkPropagator.scala

```diff
@@ -0,0 +1,299 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming
+
+import java.{util => jutil}
+
+import scala.collection.mutable
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.util.Utils
+
+/**
+ * Interface for propagating watermark. The implementation is not required to be thread-safe,
+ * as all methods are expected to be called from the query execution thread.
+ * (The guarantee may change on further improvements on Structured Streaming - update
+ * implementations if we change the guarantee.)
+ */
+sealed trait WatermarkPropagator {
```

Comment:
So what is the relationship between this and the WatermarkTracker?  The WatermarkTracker is still responsible to determining the global watermark?  There is also a map in WatermarkTracker i.e. "operatorToWatermarkMap" that is used to track the watermark on a per operator basis and is used to determine the global watermark.  Wouldn't the calculation for global watermark change if we are changing how watermarks for operators are calculated?

Reply:
The "operators" in operatorToWatermarkMap in WatermarkTracker are watermark nodes. We never had the concept of watermark per operator.

We don't change anything about calculating global watermark. We just start to disallow redefining the watermark which was allowed and root cause of incorrect global watermark value. Once global watermark is calculated from WatermarkTracker, it is considered as origin of watermark for WatermarkPropagator. Nothing should be changed for WatermarkTracker.